bolt/lib/Passes/IdenticalCodeFolding.cpp

```diff
@@ -340,6 +354,77 @@ typedef std::unordered_map<BinaryFunction *, std::vector<BinaryFunction *>,
 
 namespace llvm {
 namespace bolt {
+void IdenticalCodeFolding::initVTableReferences(const BinaryContext &BC) {
+  initVtable();
+  for (const auto &[Address, Data] : BC.getBinaryData()) {
+    // Filter out all symbols that are not vtables.
+    if (!Data->getName().starts_with("_ZTV"))
+      continue;
+    for (uint64_t I = Address, End = I + Data->getSize(); I < End; I += 8)
+      setAddressUsedInVTable(I);
+  }
+}
+void IdenticalCodeFolding::analyzeDataRelocations(BinaryContext &BC) {
+  initVTableReferences(BC);
+  for (const BinarySection &Sec : BC.sections()) {
+    if (!Sec.hasSectionRef() || !Sec.isData())
+      continue;
+    for (const auto &Rel : Sec.relocations()) {
+      const uint64_t RelAddr = Rel.Offset + Sec.getAddress();
+      if (isAddressInVTable(RelAddr))
+        continue;
+      BinaryFunction *BF = BC.getFunctionForSymbol(Rel.Symbol);
+      if (!BF)
+        continue;
+      BF->setHasAddressTaken(true);
+    }
+    for (const auto &Rel : Sec.dynamicRelocations()) {
+      const uint64_t RelAddr = Rel.Offset + Sec.getAddress();
+      if (isAddressInVTable(RelAddr))
+        continue;
+      BinaryFunction *BF =
+          BC.getBinaryFunctionContainingAddress(Rel.Addend,
+                                                /*CheckPastEnd*/ false,
+                                                /*UseMaxSize*/ true);
+      if (!BF)
+        continue;
+      BF->setHasAddressTaken(true);
+    }
+  }
+}
+void IdenticalCodeFolding::analyzeFunctions(BinaryContext &BC) {
+  ParallelUtilities::WorkFuncTy WorkFun = [&](BinaryFunction &BF) {
+    for (const BinaryBasicBlock *BB : BF.getLayout().blocks())
```

Comment:
I don't see if you need to iterate over blocks in layout order
```suggestion
    for (const BinaryBasicBlock &BB : BF)
```